module/counters/persistent_strict_monotonic_counter.go

```diff
@@ -0,0 +1,63 @@
+package counters
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/storage"
+)
+
+// PersistentStrictMonotonicCounter represents the consumer progress with strict monotonic counter.
+type PersistentStrictMonotonicCounter struct {
+	consumerProgress storage.ConsumerProgress
+
+	// used to skip heights that are lower than the current height
+	counter StrictMonotonousCounter
+}
+
+// NewPersistentStrictMonotonicCounter creates a new PersistentStrictMonotonicCounter which inserts the default
+// processed index to the storage layer and creates new counter with defaultIndex value.
+// The consumer progress and associated db entry must not be accessed outside of calls to the returned object,
+// otherwise the state may become inconsistent.
+//
+// No errors are expected during normal operation.
+func NewPersistentStrictMonotonicCounter(consumerProgress storage.ConsumerProgress, defaultIndex uint64) (*PersistentStrictMonotonicCounter, error) {
+	m := &PersistentStrictMonotonicCounter{
+		consumerProgress: consumerProgress,
+	}
+
+	// sync with storage for the processed index to ensure the consistency
+	value, err := m.consumerProgress.ProcessedIndex()
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			err := m.consumerProgress.InitProcessedIndex(defaultIndex)
+			if err != nil {
+				return nil, fmt.Errorf("could not init consumer progress: %w", err)
+			}
+			m.counter = NewMonotonousCounter(defaultIndex)
+		} else {
+			return nil, fmt.Errorf("could not read consumer progress: %w", err)
+		}
+	} else {
+		m.counter = NewMonotonousCounter(value)
+	}
+
+	return m, nil
```

Comment:
nit: flatten the logic so its easier to read
```suggestion
	if err != nil {
		if !errors.Is(err, storage.ErrNotFound) {
			return nil, fmt.Errorf("could not read consumer progress: %w", err)
		}

		err := m.consumerProgress.InitProcessedIndex(defaultIndex)
		if err != nil {
			return nil, fmt.Errorf("could not init consumer progress: %w", err)
		}

		value = defaultIndex
	}
	
	m.counter = NewMonotonousCounter(value)
	return m, nil
```

Reply:
[982af35](https://github.com/onflow/flow-go/pull/5752/commits/982af35f285f6ab761e5df32874b6631e7931ebd)